src/cascadia/UIMarkdown/MarkdownToXaml.cpp

```diff
@@ -0,0 +1,451 @@
+{
+    _EndSpan();
+    _lastParagraph = nullptr;
+}
+
+WUX::Controls::TextBlock MarkdownToXaml::_makeDefaultTextBlock()
+{
+    WUX::Controls::TextBlock b{};
+    b.IsTextSelectionEnabled(true);
+    b.TextWrapping(WUX::TextWrapping::WrapWholeWords);
+    return b;
+}
+
+void MarkdownToXaml::_RenderNode(cmark_node* node, cmark_event_type ev_type)
+{
+    cmark_node* parent;
+    cmark_node* grandparent;
+    bool tight;
+
+    bool entering = (ev_type == CMARK_EVENT_ENTER);
+
+    switch (cmark_node_get_type(node))
+    {
+    case CMARK_NODE_DOCUMENT:
+        break;
+
+    case CMARK_NODE_BLOCK_QUOTE:
+
+        // It's non-trivial to deal with the right-side vertical lines that
+        // we're accustomed to seeing for block quotes in markdown content.
+        // RichTextBlock doesn't have a good way of adding a border to a
+        // paragraph, it would seem.
+        //
+        // We could add a InlineUIContainer, with a Border in there, then
+        // put a new RichTextBlock in there, but I believe text selection
+        // wouldn't transit across the border.
+
+        // Instead, we're just going to add a new layer of indenting.
+
+        if (entering)
+        {
+            _EndParagraph();
+            _indent++;
+            _blockQuoteDepth++;
+        }
+        else
+        {
+            _EndParagraph();
+            _indent = std::max(0, _indent - 1);
+            _blockQuoteDepth = std::max(0, _blockQuoteDepth - 1);
+        }
+
+        break;
+
+    case CMARK_NODE_LIST:
+    {
+        // when `node->as.list.list_type == CMARK_BULLET_LIST`, we're an unordered list.
+        // Otherwise, we're an ordered one (and we might not start at 0).
+        //
+        // However, we don't support numbered lists for now.
+        if (entering)
+        {
+            _EndParagraph();
+            _indent++;
+        }
+        else
+        {
+            _EndParagraph();
+            _indent = std::max(0, _indent - 1);
+        }
+        break;
+    }
+
+    case CMARK_NODE_ITEM:
+        // A list item, either for a ordered list or an unordered one.
+        if (entering)
+        {
+            _EndParagraph();
+            _CurrentParagraph();
+            _NewRun().Text(winrt::hstring{ bullets[std::clamp(_indent - _blockQuoteDepth - 1, 0, 2)] });
+        }
+        break;
+
+    case CMARK_NODE_HEADING:
+        _EndParagraph();
+
+        // At the start of a header, change the font size to match the new
+        // level of header we're at. The text will come later, in a
+        // CMARK_NODE_TEXT
+        if (entering)
+        {
+            const auto level = cmark_node_get_heading_level(node);
+            _CurrentParagraph().FontSize(std::max(16u, 36u - ((level - 1) * 6u)));
+        }
+        break;
+
+    case CMARK_NODE_CODE_BLOCK:
+    {
+        _EndParagraph();
+        std::string_view code{ cmark_node_get_literal(node) };
+        const auto codeHstring{ winrt::hstring{ til::u8u16(code) } };

```

Comment:
Actually, it'd be nice if we could stop using `til::u8u16` in WT for `HSTRING`s. altogether. `winrt::to_hstring` can also convert narrow to wide strings, but does it a lot more efficiently. It's also easier to read IMO.

Reply:
winrt::to_hstring is more efficient?? that's news to me!

Reply:
I wrote that with my gut feeling and so I ran a test:
```cpp
__declspec(noinline)
static winrt::hstring winrt_to_hstring(const std::string_view& str) {
    return winrt::to_hstring(str);
}

__declspec(noinline)
static winrt::hstring u8u16_hstring(const std::string_view& str) {
    return winrt::hstring{ til::u8u16(str) };
}

static void BM_winrt_to_hstring(benchmark::State& state) {
    for (auto _ : state) {
        auto h = winrt_to_hstring("Hello, World!");
        benchmark::DoNotOptimize(h);
    }
}
BENCHMARK(BM_winrt_to_hstring);

static void BM_u8u16_hstring(benchmark::State& state) {
    for (auto _ : state) {
        auto h = u8u16_hstring("Hello, World!");
        benchmark::DoNotOptimize(h);
    }
}
BENCHMARK(BM_u8u16_hstring);

BENCHMARK_MAIN();
```

Result:
```
--------------------------------------------------------------
Benchmark                    Time             CPU   Iterations
--------------------------------------------------------------
BM_winrt_to_hstring       52.4 ns         52.3 ns    135757576
BM_u8u16_hstring          80.9 ns         80.7 ns     86153846
```

Reply:
then why do we have `u8u16`???? I thought the whole point was to be faster than `winrt::to_hstring`!

Reply:
As far as I remember it was introduced to solve the issue of codepoints being split across two writes (`u16state` / `u16state`). I don't think using it in WT was ever an explicit goal, and I believe that just happened coincidentally. (#4093 introduced those functions.)

FWIW it also doesn't get any better if the strings are longer than "Hello, World!", because the second `MultiByteToWideChar` call that `u8u16` saves is very cheap, while it uses `string.resize(capacity)` which fills the string will null bytes (= O(n), similarly expensive).

Reply:
It's probably worth noting that `til::u8u16` is more expensive than `winrt::to_hstring` across all string lengths, even if the code is changed to return the `wstring` directly without copying it to a HSTRING first:
```cpp
static std::wstring u8u16_hstring(const std::string_view& str) {
    return til::u8u16(str);
}
```

This happens because `til::u8u16` wraps a `noexcept` overload of itself, which has a try/catch clause which prevents proper inlining. Despite the C++ exception dogma, they're generally anything but zero-cost, particularly try/catch clauses.